contracts/ReverseRegistrar.sol

```diff
@@ -1,31 +1,117 @@
+        name[node] = _name;
+    }
+}
 
 contract ReverseRegistrar {
+    // namehash('addr.reverse')
+    bytes32 constant ADDR_REVERSE_NODE = 0x91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e2;
+
     AbstractENS public ens;
-    bytes32 public rootNode;
+    Resolver public defaultResolver;
 
     /**
      * @dev Constructor
      * @param ensAddr The address of the ENS registry.
-     * @param node The node hash that this registrar governs.
+     * @param _resolver The address of the default reverse resolver.
      */
-    function ReverseRegistrar(AbstractENS ensAddr, bytes32 node) {
+    function ReverseRegistrar(AbstractENS ensAddr, Resolver _resolver) {
         ens = ensAddr;
-        rootNode = node;
+        defaultResolver = _resolver;
     }
-
+    
     /**
      * @dev Transfers ownership of the reverse ENS record associated with the
      *      calling account.
      * @param owner The address to set as the owner of the reverse record in ENS.
      * @return The ENS node hash of the reverse record.
      */
     function claim(address owner) returns (bytes32 node) {
+        return claimWithResolver(owner, 0);
+    }
+
+    /**
+     * @dev Transfers ownership of the reverse ENS record associated with the
+     *      calling account.
+     * @param owner The address to set as the owner of the reverse record in ENS.
+     * @param resolver The address of the resolver to set; 0 to leave unchanged.
+     * @return The ENS node hash of the reverse record.
+     */
+    function claimWithResolver(address owner, address resolver) returns (bytes32 node) {
         var label = sha3HexAddress(msg.sender);
-        ens.setSubnodeOwner(rootNode, label, owner);
-        return sha3(rootNode, label);
+        node = sha3(ADDR_REVERSE_NODE, label);
+        var currentOwner = ens.owner(node);
+
+        // Update the resolver if required
+        if(resolver != 0 && resolver != ens.resolver(node)) {
+            // Transfer the name to us first if it's not already
+            if(currentOwner != address(this)) {
+                ens.setSubnodeOwner(ADDR_REVERSE_NODE, label, this);
+                currentOwner = address(this);
+            }
+            ens.setResolver(node, resolver);
+        }
+
+        // Update the owner if required
+        if(currentOwner != owner) {
+            ens.setSubnodeOwner(ADDR_REVERSE_NODE, label, owner);
+        }
+
+        return node;
+    }
+
+    /**
+     * @dev Sets the `name()` record for the reverse ENS record associated with
+     * the calling account. First updates the resolver to the default reverse
+     * resolver if necessary.
+     * @param name The name to set for this address.
+     * @return The ENS node hash of the reverse record.
+     */
+    function setName(string name) returns (bytes32 node) {
+        node = claimWithResolver(this, defaultResolver);
+        defaultResolver.setName(node, name);
```

Comment:
It doesn't - you can set your reverse record to any name; callers have to check that it's consistent. Forcing this here wouldn't help, because someone could set the name to match, then update it later - or just set their own resolver that behaves however they want.